File: python/0141_Linked_List_Cycle.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        if not head or not head.next:
            return False
        slow = head
        fast = head.next
        while slow and fast and fast.next:
            if fast is slow or fast.next is slow:
                return True
            slow = slow.next
            fast = fast.next.next
        return False

# O(n) Time
# O(1) Space

File: cpp/0141_Linked_List_Cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (!head || !(head->next)) return false;
        ListNode *slow = head, *fast = head->next;
        while (fast && fast->next) {
            if (fast == slow || fast->next == slow) return true;
            slow = slow->next;
            fast = fast->next->next;
        }
        return false;
    }
};

// O(N) Time
// O(1) Space